apps/labs/posts/multi-dimensional-sparse-arrays.md

```diff
@@ -0,0 +1,225 @@
+
+  2.  **Tensordot of a sparse array with a dense array** -
+      This implementation is also similar to `dot`'s, but involves some additional processing.
+      Unlike `dot`, where the last dimension is always the reduced one, `tensordot` can take a tuple of `axes` to be reduced. Thus, we cannot directly use `dot`. Instead, we accumulate all reduced axes into one dimension and make this the trailing (rightmost) dimension. The approach taken for this involved transposing the array with a permutation based on the `axes` argument, so that all the reduced axes became the trailing axes of the transposed array. This was followed by a reshape, where the non-reduced axes remained as they were, but the reduced ones were contracted into a single dimension with a size equal to the product of the dimension sizes of the reduced axes.
+
+      This was definitely one of my favourite ideas that I came up with and worked around - the transposing and permuting involved lots of experimentation with NumPy arrays, until it all fell into place!
+
+This concluded the work for the second PR. Moving on to the third stage, I focused on implementing element-wise operations such as `multiply`, `divide`, `minimum`, `maximum`, all boolean comparators (`==`, `!=`, `>`, `<`, `>=`, `<=`) etc. The pre-existing behaviour of these was based on a straightforward logic - two arrays could only undergo element-wise operations if they had the same shape. Most of these operations converted the input arrays to CSR/CSC format before further computation. However, at that time, 2-D CSR broadcasting didn't exist, and the need for this to eventually be introduced went all the way back to [this issue](https://github.com/scipy/scipy/issues/2128) from 2013. The goal was to ensure that the behaviour of methods in SciPy sparse replicated that of the corresponding NumPy methods.
+
+The work for this PR, therefore, started off by opening a PR to [add CSR broadcasting](https://github.com/scipy/scipy/pull/21412). This followed by making a number of tweaks to the pre-existing CSR/CSC methods, which included removing outdated tests which no longer raised an error, now that 1D and 2D CSR arrays could be multiplied without resulting in a `ValueError`, provided they were broadcastable. Ensuring that CSR arrays could broadcast before making changes to the COO codebase was essential for maintaining consistency in behaviour across all sparse array formats.
+
+Since COO broadcasting had already been implemented in the second PR, extending most element-wise operations to n-D followed a standarrd procedure. This involved broadcasting the input arrays, reshaping them to 2-D, routing via CSR, and then reshaping the output back. This update also necessitated modifying the definitions of addition and subtraction to incorporate broadcasting. This led to a change in the definition of densification for COO arrays (performed by the constructor `toarray()`, one of the first methods defined).
```

Comment:
```suggestion
Since COO broadcasting had already been implemented in the second PR, extending most element-wise operations to n-D followed a standard procedure. This involved broadcasting the input arrays, reshaping them to 2-D, routing via CSR, and then reshaping the output back. This update also necessitated modifying the definitions of addition and subtraction to incorporate broadcasting. This led to a change in the definition of densification for COO arrays (performed by the constructor `toarray()`, one of the first methods defined).
```